Source/WTF/wtf/text/StringImpl.h

```diff
@@ -273,13 +272,14 @@ class StringImpl : private StringImplShape {
     static Ref<StringImpl> createByReplacingInCharacters(const LChar*, unsigned length, UChar target, UChar replacement, unsigned indexOfFirstTargetCharacter);
     static Ref<StringImpl> createByReplacingInCharacters(const UChar*, unsigned length, UChar target, UChar replacement, unsigned indexOfFirstTargetCharacter);
 
-    static Ref<StringImpl> createStaticStringImpl(const char* characters, unsigned length)
+    static Ref<StringImpl> createStaticStringImpl(std::span<const char> characters)
     {
-        ASSERT(charactersAreAllASCII(std::span { bitwise_cast<const LChar*>(characters), static_cast<size_t>(length) }));
-        return createStaticStringImpl(bitwise_cast<const LChar*>(characters), length);
+        std::span lcharSpan { reinterpret_cast<const LChar*>(characters.data()), characters.size() };
```

Comment:
Not `spanReinterpretCast`? Is it a “too many includes” issue?

Reply:
That's right, this header doesn't currently include wtf/Algorithm.h and I was worried about increase build times. That said, I've been thinking that maybe `spanReinterpretCast` should be in wtf/StdLibExtras.h instead, which *is* included in Stringimpl.h.

Reply:
Oh now that I made the change and let EWS chew on it, I remember why I didn't do this yet...

```
FAILED: Source/ThirdParty/skia/CMakeFiles/Skia.dir/__/__/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp.o 
/usr/bin/ccache /usr/bin/c++ -DBUILDING_WEBKIT=1 -DBUILDING_WITH_CMAKE=1 -DBUILDING_WPE__=1 -DBWRAP_EXECUTABLE=\"/usr/bin/bwrap\" -DDBUS_PROXY_EXECUTABLE=\"/usr/bin/xdg-dbus-proxy\" -DGETTEXT_PACKAGE=\"WPE\" -DHAVE_CONFIG_H=1 -DJSC_GLIB_API_ENABLED -DPAS_BMALLOC=1 -DSKIA_IMPLEMENTATION=1 -DSK_ASSUME_GL=0 -DSK_ASSUME_GL_ES=1 -DSK_ASSUME_WEBGL=0 -DSK_CODEC_DECODES_PNG -DSK_DISABLE_LEGACY_GL_MAKE_NATIVE_INTERFACE -DSK_DISABLE_LEGACY_IMAGE_READBUFFER -DSK_DISABLE_LEGACY_INIT_DECODERS -DSK_DISABLE_LEGACY_PNG_WRITEBUFFER -DSK_DISABLE_TRACING -DSK_DISABLE_WEBGL_INTERFACE -DSK_ENABLE_AVX512_OPTS -DSK_ENABLE_PRECOMPILE -DSK_GAMMA_APPLY_TO_A8 -DSK_GANESH -DSK_GL -DSK_R32_SHIFT=16 -DSK_RELEASE -DSK_TYPEFACE_FACTORY_FREETYPE -DSTATICALLY_LINKED_WITH_WTF -DSTATICALLY_LINKED_WITH_bmalloc -I/app/webkit/Source/WTF -I/app/webkit/WebKitBuild/WPE/Release -I/app/webkit/Source/ThirdParty/skia -I/app/webkit/WebKitBuild/WPE/Release/WTF/Headers -I/app/webkit/WebKitBuild/WPE/Release/bmalloc/Headers -isystem /usr/include/freetype2 -fdiagnostics-color=always -Wextra -Wall -pipe -fmax-errors=20 -Wno-odr -Wno-stringop-overread -Wno-stringop-overflow -Wno-nonnull -Wno-array-bounds -Wno-expansion-to-defined -Wno-noexcept-type -Wno-psabi -Wno-misleading-indentation -Wno-maybe-uninitialized -Wundef -Wpointer-arith -Wmissing-format-attribute -Wformat-security -Wcast-align -Wno-tautological-compare -fno-strict-aliasing -fno-exceptions -fno-rtti -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -ffp-contract=off -fstrict-aliasing -Wno-attributes -Wno-cast-align -Wno-dangling-reference -Wno-deprecated -Wno-psabi -Wno-suggest-attribute=format -Wno-undef -Wno-uninitialized -Wno-unused-parameter -MD -MT Source/ThirdParty/skia/CMakeFiles/Skia.dir/__/__/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp.o -MF Source/ThirdParty/skia/CMakeFiles/Skia.dir/__/__/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp.o.d -o Source/ThirdParty/skia/CMakeFiles/Skia.dir/__/__/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp.o -c /app/webkit/Source/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp
In file included from /app/webkit/Source/WTF/wtf/FastMalloc.h:26,
                 from /app/webkit/Source/WTF/config.h:31,
                 from /app/webkit/Source/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp:26:
/app/webkit/Source/WTF/wtf/StdLibExtras.h:675:6: error: ‘span’ in namespace ‘std’ does not name a template type
  675 | std::span<T> spanReinterpretCast(std::span<U> span)
      |      ^~~~
/app/webkit/Source/WTF/wtf/StdLibExtras.h:675:1: note: ‘std::span’ is only available from C++20 onwards
  675 | std::span<T> spanReinterpretCast(std::span<U> span)
      | ^~~
/app/webkit/Source/WTF/wtf/StdLibExtras.h:733:12: error: ‘spanReinterpretCast’ has not been declared in ‘WTF’
```

The Skia bot is importing this header in some of their Skia logic and they're not building with C++20 support enabled (so no std::span support) :(

Reply:
We must fix the build so it compiles those source files with C++20 support enabled; not acceptable to compile our headers without it

Reply:
Is that challenging?

Reply:
I have made people maintaining the skia port aware of the issue. I do agree they should build with C++20 support enabled. In the mean time, I did address the bot failure by adding a C++ version guard around the new function in the header.

Reply:
Oh, it looks like it is still broken for Igalia's JSC bots though :(
```
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h: In static member function ‘static WTF::Ref<WTF::StringImpl> WTF::StringImpl::create(std::span<const char>)’:
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h:250:99: error: ‘spanReinterpretCast’ was not declared in this scope
  250 |     ALWAYS_INLINE static Ref<StringImpl> create(std::span<const char> characters) { return create(spanReinterpretCast<const LChar>(characters)); }
      |                                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h: In static member function ‘static WTF::Ref<WTF::StringImpl> WTF::StringImpl::createWithoutCopying(std::span<const char>)’:
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h:266:127: error: ‘spanReinterpretCast’ was not declared in this scope
  266 |     ALWAYS_INLINE static Ref<StringImpl> createWithoutCopying(std::span<const char> characters) { return createWithoutCopying(spanReinterpretCast<const LChar>(characters)); }
      |                                                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h: In static member function ‘static WTF::Ref<WTF::StringImpl> WTF::StringImpl::createStaticStringImpl(std::span<const char>)’:
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h:277:31: error: ‘spanReinterpretCast’ was not declared in this scope
  277 |         std::span lcharSpan = spanReinterpretCast<const LChar>(characters);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
gmake[3]: *** [Source/WTF/wtf/CMakeFiles/WTF.dir/build.make:186: Source/WTF/wtf/CMakeFiles/WTF.dir/CodePtr.cpp.o] Error 1
```

Reply:
I ping'd Igalia about this as well. But due to timezone differences, I may need to go back to my original version until theirs ports consistently build with C++20 support enabled (it was not directly obvious to me how to make that change for their ports).

Reply:
`#if __cplusplus >= 202002L` doesn't evaluate to true on those Igalia JSC bots but they can build span code though...